### PR TITLE
vm-pool, Reject VMs with duplicate interfaces

### PR DESCRIPTION
**Pull request by RamLavi**
**What this PR does / why we need it**:
When applying a VM with multiple interfaces with the same name,
the VM is rejected but kubemacpool is mishandling the MACs,
causing a side effect where that MAC is no longer usable (it
is "taken" by the ghost VM).
This commit fixes this by checking and rejecting VMs with
duplicate interface names on kmp webhook context.

**Special notes for your reviewer**:

**Release note**:

```release-note
Kubemacpool webhook behavior is changed: virtualMachine with duplicate interface names will be rejected from now on.
```

### CODE REVIEW

**Comment by RamLavi** [comment]
/hold
need to check pods module as well

**Comment by RamLavi** [comment]
/hold cancel
pods are not relevant for this issue

File: pkg/pool-manager/virtualmachine_pool.go [hunk 1]
```diff
@@ -199,6 +213,19 @@ func getVirtualMachineNetworks(virtualMachine *kubevirt.VirtualMachine) []kubevi
...
+
+	tmpInterfaceMap := map[string]string{}
+	for _, vmInterface := range vmInterfaces {
+		if _, exists := tmpInterfaceMap[vmInterface.Name]; exists {
+			return false
+		}
+		tmpInterfaceMap[vmInterface.Name] = ""
+	}
+	return true
+}
```
Comment (qinqon) [inline pkg/pool-manager/virtualmachine_pool.go:227]:
```suggestion
func checkVmForInterfaceDuplication(virtualMachine *kubevirt.VirtualMachine) error {
	vmInterfaces := getVirtualMachineInterfaces(virtualMachine)

	tmpInterfaceMap := map[string]string{}
	for _, vmInterface := range vmInterfaces {
		if _, exists := tmpInterfaceMap[vmInterface.Name]; exists {
			return fmt.Errorf("Failed to mutate virtual machine: every interface must have a unique name")
		}
		tmpInterfaceMap[vmInterface.Name] = ""
	}
	return nil
}
```

This way on the two places we call it we don't have to duplicate the error

```golang
if err := checkVmForInterfaceDuplication(virtualMachine) ; err != nil {
    return err
}
		}
```

**Review by qinqon — CHANGES_REQUESTED** [review]

**Review by maiqueb — COMMENTED** [review]
Some comments.

File: pkg/pool-manager/virtualmachine_pool.go (hunk 1 above)
Comment (RamLavi) [reply pkg/pool-manager/virtualmachine_pool.go:227]: Done

File: pkg/pool-manager/virtualmachine_pool.go [hunk 2]
```diff
@@ -47,6 +47,13 @@ func (p *PoolManager) AllocateVirtualMachineMac(virtualMachine *kubevirt.Virtual
 		return nil
 	}
 
+	if isNotDryRun {
+		// We can't allow for duplicate interfaces names, as interface.Name is macMap's key.
+		if !checkVmForInterfaceDuplication(virtualMachine) {
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply pkg/pool-manager/virtualmachine_pool.go:52]: Done

File: pkg/pool-manager/virtualmachine_pool.go [hunk 3]
```diff
@@ -126,6 +133,13 @@ func (p *PoolManager) UpdateMacAddressesForVirtualMachine(previousVirtualMachine
 	}
 	defer p.poolMutex.Unlock()
 
+	if isNotDryRun {
+		// We can't allow for duplicate interfaces names, as interface.Name is macMap's key.
+		if !checkVmForInterfaceDuplication(virtualMachine) {
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply pkg/pool-manager/virtualmachine_pool.go:138]: Done

File: pkg/pool-manager/virtualmachine_pool.go [hunk 4]
```diff
@@ -199,6 +213,19 @@ func getVirtualMachineNetworks(virtualMachine *kubevirt.VirtualMachine) []kubevi
 	return virtualMachine.Spec.Template.Spec.Networks
 }
 
+func checkVmForInterfaceDuplication(virtualMachine *kubevirt.VirtualMachine) bool {
+	vmInterfaces := getVirtualMachineInterfaces(virtualMachine)
+
+	tmpInterfaceMap := map[string]string{}
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply pkg/pool-manager/virtualmachine_pool.go]: Done

File: pkg/pool-manager/virtualmachine_pool.go [hunk 5]
```diff
@@ -47,6 +47,11 @@ func (p *PoolManager) AllocateVirtualMachineMac(virtualMachine *kubevirt.Virtual
 		return nil
 	}
 
+	// We can't allow for duplicate interfaces names, as interface.Name is macMap's key.
+	if err := checkVmForInterfaceDuplication(virtualMachine); isNotDryRun && err != nil {
```
Comment (qinqon) [inline pkg/pool-manager/virtualmachine_pool.go:51]: We don't want to even run `checkVmForInterfaceDuplication` if we are on dry mode right ?

**Review by qinqon — CHANGES_REQUESTED** [review]

File: pkg/pool-manager/virtualmachine_pool.go (hunk 5 above)
Comment (maiqueb) [reply pkg/pool-manager/virtualmachine_pool.go:51]: yeah, with inlining I meant that we should do something like `if !isDryRun && ...`

Comment (qinqon) [reply pkg/pool-manager/virtualmachine_pool.go:51]: @RamLavi can you fix it ?

Comment (RamLavi) [reply pkg/pool-manager/virtualmachine_pool.go:51]: I don't think it's a working syntax. if we want to not run `checkVmForInterfaceDuplication` if not needed  then either we do but don't return the err from `checkVmForInterfaceDuplication`:
```
if isNotDryRun && checkVmForInterfaceDuplication(virtualMachine) != nil {
    ...
}
```
or do  it in a 2 lines like initially done
```
if isNotDryRun {
    if err := checkVmForInterfaceDuplication(virtualMachine); err != nil {
        return err
    }
}
```
do you have another option?

Comment (maiqueb) [reply pkg/pool-manager/virtualmachine_pool.go]: Yes, the latter (without in-lining) seems preferable. Sorry for dragging you into this.

Comment (RamLavi) [reply pkg/pool-manager/virtualmachine_pool.go]: no worries, better it be as clean as possible :) 

even better, moved the flag inside the function on @qinqon 's suggestion. what do you think?

**Review by qinqon — APPROVED** [review]
/lgtm

File: pkg/pool-manager/virtualmachine_pool.go (hunk 5 above)
Comment (maiqueb) [reply pkg/pool-manager/virtualmachine_pool.go]: I think boolean arguments suck. The caller knows what to do: call the function, or avoid doing so, based on the value of the boolean.

No point passing it around. 

My 2 cents :)

Comment (qinqon) [reply pkg/pool-manager/virtualmachine_pool.go]: Let's not inline the boolean check then and do as @RamLavi proposed

Comment (RamLavi) [reply pkg/pool-manager/virtualmachine_pool.go]: [Done](https://github.com/k8snetworkplumbingwg/kubemacpool/compare/1ea79aa610967f9f4b42500bcfc7377e7ed1dafc..53f7cda95cc5b39942d4b6662a256682ec13ff49)

**Comment by maiqueb** [comment]
I wonder about the approach altogether; afaiu, how the mutating webhook handles failures depend on its configuration - specifically, the value of its [`failurePolicy`](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#failure-policy). 

IIUC, this approach will only work if the failure policy is `fail`; since we own the webhook, that **should not be an issue**. Never the less, this might deserve a call out in the release notes: this approach is tied to the `fail` failure policy - which is the mutating webhook's default.

**Comment by maiqueb** [comment]
> I wonder about the approach altogether; afaiu, how the mutating webhook handles failures depend on its configuration - specifically, the value of its [`failurePolicy`](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#failure-policy).
> 
> IIUC, this approach will only work if the failure policy is `fail`; since we own the webhook, that **should not be an issue**. Never the less, this might deserve a call out in the release notes: this approach is tied to the `fail` failure policy - which is the mutating webhook's default.

/hold 

The release notes should be updated indicating what is being fixed, and under what conditions - i.e. the `MutatingWebhookConfiguration` cannot be updated / must feature the `fail` failure policy.

**Comment by RamLavi** [comment]
> I wonder about the approach altogether; afaiu, how the mutating webhook handles failures depend on its configuration - specifically, the value of its [`failurePolicy`](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#failure-policy).
> 
> IIUC, this approach will only work if the failure policy is `fail`; since we own the webhook, that **should not be an issue**. Never the less, this might deserve a call out in the release notes: this approach is tied to the `fail` failure policy - which is the mutating webhook's default.

Well this webhook implementation has always relied on the fact that the `failurePolicy` is `fail`. nothing changed in that regard. Moreover, the check we are adding is already being done, only later, in the webhook validation phase.
So, considering what you say, we have 2 options here:
1. we can fail the VM on our webhook, saying we cannot deal with duplicate interface names, due to our current implementation (this is what we do in this PR)
2. we can ignore the VM on our webhook (without assigning MACs, thus avoiding the side effect), letting the VM fail in the validation part later on. 

IMO I'm not too excited about option 2. I prefer to fail instead of ignoring the issue.. 
what do you think @maiqueb @qinqon ?

**Comment by qinqon** [comment]
> > I wonder about the approach altogether; afaiu, how the mutating webhook handles failures depend on its configuration - specifically, the value of its [`failurePolicy`](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#failure-policy).
> > IIUC, this approach will only work if the failure policy is `fail`; since we own the webhook, that **should not be an issue**. Never the less, this might deserve a call out in the release notes: this approach is tied to the `fail` failure policy - which is the mutating webhook's default.
> 
> Well this webhook implementation has always relied on the fact that the `failurePolicy` is `fail`. nothing changed in that regard. Moreover, the check we are adding is already being done, only later, in the webhook validation phase. So, considering what you say, we have 2 options here:
> 
> 1. we can fail the VM on our webhook, saying we cannot deal with duplicate interface names, due to our current implementation (this is what we do in this PR)
> 2. we can ignore the VM on our webhook (without assigning MACs, thus avoiding the side effect), letting the VM fail in the validation part later on.
> 
> IMO I'm not too excited about option 2. I prefer to fail instead of ignoring the issue.. what do you think @maiqueb @qinqon ?

Fail fast is always better.

**Comment by maiqueb** [comment]
> > > I wonder about the approach altogether; afaiu, how the mutating webhook handles failures depend on its configuration - specifically, the value of its [`failurePolicy`](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#failure-policy).
> > > IIUC, this approach will only work if the failure policy is `fail`; since we own the webhook, that **should not be an issue**. Never the less, this might deserve a call out in the release notes: this approach is tied to the `fail` failure policy - which is the mutating webhook's default.
> > 
> > 
> > Well this webhook implementation has always relied on the fact that the `failurePolicy` is `fail`. nothing changed in that regard. Moreover, the check we are adding is already being done, only later, in the webhook validation phase. So, considering what you say, we have 2 options here:

Right. Which means the mutating webhook mutates, and the validating webhook validates. So far so good. 

> > 
> > 1. we can fail the VM on our webhook, saying we cannot deal with duplicate interface names, due to our current implementation (this is what we do in this PR)

This is making the mutating webhook validate. Which is something that - imo - belongs on the validating webhook. I won't pretend that I am versed enough in webhook design patterns to claim that validating stuff in the mutating webhook is an anti-pattern, but it surely feels semantically weird. That's why I advocated out of band for the reconcile loop to sort this inconsistency.

> > 2. we can ignore the VM on our webhook (without assigning MACs, thus avoiding the side effect), letting the VM fail in the validation part later on.

And let eventually the reconcile loop correct the state. IMO, this is more aligned with the kubernetes design model.
> > 
> > IMO I'm not too excited about option 2. I prefer to fail instead of ignoring the issue.. what do you think @maiqueb @qinqon ?
> 
> Fail fast is always better.

...
Now the key thing is (quoting @RamLavi) :
> > Well this webhook implementation has always relied on the fact that the `failurePolicy` is `fail`. nothing changed in that regard. 

Right, meaning whatever implicit rules I'm trying to uphold were already broken. No point arguing about them now; your proposed fix seems more in-line with the current design of this component.

Please detail the fix in the release notes - afaiu, you're changing the behavior of the webhook. Furthermore ... shouldn't you also delete this particular validation from the validating webhook ? It will be redundant now.

**Review by maiqueb — APPROVED** [review]
Thank you for your patience.

**Pull request merged**